In [ ]:
import pandas as pd
df = pd.read_csv("/media/sagark24/New Volume/MERGE CDIS/IPYNB_FILE/DATA_GENERATION/SWNO_MASTER_COMBINED_FULL.csv", dtype=str)
col = df['SOURCE_SS'].unique()
print(col)

['220 kV AAREY EHV STATION' 'AAREY 220 KV RECEIVING STATION'
 '220kV SAKI REC-STN' '220kV GOREGAON REC-STN' '220kV VERSOVA REC-STN'
 'MAHANANDA REC-STN' 'NESCO RECEIVING STATION' 'SAHAR PLAZA REC-STN'
 "TATA'S DHARAVI REC-STN" "TATA'S MALAD REC-STN" "TATA'S BORIVALI REC-STN"
 "TATA's DHARAVI REC-STN" "TATA'S SAKI REC-STN" "TATA'S VERSOVA REC-STN"
 "TATA'S VIKHROLI REC-STN" 'AAREY 220KV R/S' 'GOKULDHAM REC-STN'
 'HINDUSTAN UNILEVER REC-STN' 'TATA BKC REC-STN' '220kV BORIVALI REC-STN'
 'DAHISAR REC-STN' 'KANAKIA CCI REC-STN' 'SHIMPOLI REC-STN'
 'BORIVALI REC-STN' '220kV CHEMBUR REC-STN' '220KV CHEMBUR REC-STN'
 "TATA'S CHEMBUR REC-STN" 'PALI REC-STN' 'BOMBILWADI REC-STN'
 'POISAR REC-STN' '220kV GHODBUNDER REC-STN' 'DAHISAR WEST REC-STN'
 'DEVIDAS LANE REC-STN' 'MIRA REC-STN' 'BHAYANDAR WEST REC-STN'
 'VINAMRA REC-STN' 'SHANTI STAR MIRA REC-STN' 'RAVI MHADA REC-STN'
 'SADGURU REC-STN' '220kV GORAI REC-STN' 'VAZIRA REC-STN' 'K.I.E. REC-STN'
 'GORAI REC-STN' 'NIRLON REC-STN' 'MAHANANDA' "T

| **Standardized Name** | **All Variants Found**                                                           | **Count** |
| --------------------- | -------------------------------------------------------------------------------- | --------- |
| AAREY 220KV R/S       | `'220 kV AAREY EHV STATION'`, `'AAREY 220KV R/S'`                                | 2         |
| SAKI REC-STN          | `'220kV SAKI REC-STN'`, `"TATA'S SAKI REC-STN"`                                  | 2         |
| TATA DHARAVI REC-STN  | `"TATA'S DHARAVI REC-STN"`, `"TATA's DHARAVI REC-STN"`                           | 2         |
| VERSOVA REC-STN       | `'220kV VERSOVA REC-STN'`, `"TATA'S VERSOVA REC-STN"`                            | 2         |
| CHEMBUR REC-STN       | `'220kV CHEMBUR REC-STN'`, `'220KV CHEMBUR REC-STN'`, `"TATA'S CHEMBUR REC-STN"` | 3         |
| BORIVALI REC-STN      | `'220kV BORIVALI REC-STN'`, `'BORIVALI REC-STN'`, `"TATA'S BORIVALI REC-STN"`    | 3         |
| MAHANANDA REC-STN     | `'MAHANANDA REC-STN'`, `'MAHANANDA'`                                             | 2         |
| GORAI REC-STN         | `'GORAI REC-STN'`, `'220kV GORAI REC-STN'`                                       | 2         |

In [ ]:
col = df['DESTINATION_SS'].unique()
len(col)
print(col)

['MAROL REC-STN' 'KALANAGAR REC-STN' 'KURLA REC-STN'
 'NIRLON B10 RECEIVING STATION' 'KOHINOOR RECEIVING STATION'
 'OSHIWARA RECEIVING STATION' 'MAROL' 'MEGHAWADI REC-STN' 'MALAD REC-STN'
 'ANDHERI REC-STN' 'VIKHROLI REC-STN' 'NESCO RECEIVING STATION'
 'NIRLON REC-STN' 'HINDUSTAN UNILEVER REC-STN' 'BOMBILWADI REC-STN'
 'K.I.E. REC-STN' "TATA'S BORIVLI REC-STN" 'MAHINDRA AND MAHINDRA REC-STN'
 'MIRA REC-STN' 'TIMES OF INDIA REC-STN' '24TH ROAD REC-STN'
 'MMRDA REC-STN' 'CHINCHAVLI REC-STN' 'PALM COURT'
 'J. B. NAGAR RECEIVING STATION' 'M.I.D.C. REC-STN' 'AMBIVLI REC-STN'
 'JUHU REC-STN' 'HINGWALA LANE REC-STN' 'KALPATARU LBS R/S'
 'TILAK NAGAR REC-STN' 'HIRANANDANI REC-STN' 'VILE-PARLE REC-STN'
 'HCC REC-STN' 'SEEPZ REC-STN' 'SAMBHAJI NAGAR REC-STN' 'NAHAR SHAKTI R/S'
 'SAHAR PLAZA REC-STN' 'SAHAR PLAZA' 'NAHAR SHAKTI REC-STN'
 'SHRADDHANAND ROAD REC-STN' 'RUNWAL REC-STN' 'SAMBHAJI NAGAR'
 'GOKULDHAM REC-STN' 'RAHEJA I.T. PARK REC-STN' 'SARVODAY NAGAR REC-STN'
 'MAHANANDA REC-STN' 'CHAK

| **Standardized Station Name** | **Variants Found**                                                                |
| ----------------------------- | --------------------------------------------------------------------------------- |
| MAROL REC-STN                 | `'MAROL'`, `'MAROL REC-STN'`                                                      |
| SAHAR PLAZA REC-STN           | `'SAHAR PLAZA'`, `'SAHAR PLAZA REC-STN'`                                          |
| MAHANANDA REC-STN             | `'MAHANANDA'`, `'MAHANANDA REC-STN'`                                              |
| CHEMBUR REC-STN               | `'CHEMBUR REC-STN'`, `'220kV CHEMBUR REC-STN'`                                    |
| BORIVALI REC-STN              | `"TATA'S BORIVLI REC-STN"`, `'BORIVALI REC-STN'`                                  |
| DAHISAR REC-STN               | `'DAHISAR REC-STN'`, `'DAHISAR WEST REC-STN'`, `'DAHISAR CHECKNAKA REC-STN'`      |
| NIRLON REC-STN                | `'NIRLON REC-STN'`, `'NIRLON B10 RECEIVING STATION'`                              |
| KANAKIA CCI REC-STN           | `'KANAKIA CCI'`, `'KANAKIA CCI REC-STN'`                                          |
| BHAYANDAR REC-STN             | `'SHANTI STAR MIRA REC-STN'`, `'BHAYANDAR REC-STN'`, `'BHAYANDAR WEST REC-STN'`   |
| SAMBHAJI NAGAR REC-STN        | `'SAMBHAJI NAGAR'`, `'SAMBHAJI NAGAR REC-STN'`                                    |
| DINDOSHI REC-STN              | `'DINDOSHI REC-STN'`, `'OMKAR RECEIVING STATION'`, `'DINDOSHI VIA OMKAR REC STN'` |
| NAHAR SHAKTI REC-STN          | `'NAHAR SHAKTI R/S'`, `'NAHAR SHAKTI REC-STN'`                                    |
| JUHU REC-STN                  | `'JUHU REC-STN'`, `'JUHU NORTH REC-STN'`                                          |
| ANDHERI REC-STN               | `'ANDHERI REC-STN'`, `'ANDHERI'`                                                  |
| AMBIVLI REC-STN               | `'AMBIVLI REC-STN'`, `'AMBIVLI'`                                                  |
| KALPATARU LBS REC-STN         | `'KALPATARU LBS R/S'`, `'KALPATARU LBS REC-STN'`                                  |
| NETMAGIC DC-9 NO 1            | `'NETMAGIC DC-9 NO 1 BUS 1 DSS'`, `'NETMAGIC DC-9 NO 1 RECEIVING STATIO'`         |
| PALM COURT REC-STN            | `'PALM COURT'`, `'PALM COURT REC-STN'`                                            |


ML ALGORITHM TO LEARN WEIGHT FROM OLD HALTH SCORE FUNCTION

In [1]:
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import StandardScaler
from datetime import datetime

# Load your CSV file
df = pd.read_csv("/media/sagark24/New Volume/MERGE CDIS/IPYNB_FILE/DATA_GENERATION/SWNO_MASTER_COMBINED_FULL.csv")  # ⬅ Replace with actual file name

# Step 1: Feature Engineering
now_year = datetime.now().year

if 'DATECREATED' in df.columns:
    years = pd.to_datetime(df['DATECREATED'], errors='coerce').dt.year
    df['CABLE_AGE'] = now_year - years.fillna(now_year)

# Current mean and std
current_cols = [c for c in df.columns if c.startswith('I_Month_')]
if current_cols:
    currents = df[current_cols].apply(pd.to_numeric, errors='coerce').fillna(0)
    df['CURRENT_MEAN'] = currents.mean(axis=1)
    df['CURRENT_STD'] = currents.std(axis=1)

# PEAK, CYCLE, OVR
for prefix, name in [('PEAK_Month_', 'PEAK'), ('CYCLE_Month_', 'CYCLE'), ('OVR_Month_', 'OVR')]:
    cols = [c for c in df.columns if c.startswith(prefix)]
    if cols:
        arr = df[cols].apply(pd.to_numeric, errors='coerce').fillna(0)
        df[f'{name}_MEAN'] = arr.mean(axis=1)
        df[f'{name}_STD'] = arr.std(axis=1)

# Cable type risk
if 'CABLETYPE' in df.columns:
    df['CABLETYPE_RISK'] = df['CABLETYPE'].apply(
        lambda t: 1 if 'PILC' in str(t).upper() else (0.2 if 'XLPE' in str(t).upper() else 0.5)
    )

# Material risk
if 'CABLECONDUCTORMATERIAL' in df.columns:
    df['MATERIAL_RISK'] = df['CABLECONDUCTORMATERIAL'].apply(
        lambda m: 0.7 if 'AL' in str(m).upper() else (0.2 if 'CU' in str(m).upper() or 'COPPER' in str(m).upper() else 0.5)
    )

# FUSE count
fuse_cols = [col for col in df.columns if col.startswith('FAULT_RELAY_FUSE')]
if fuse_cols:
    df['FUSE_COUNT'] = df[fuse_cols].apply(pd.to_numeric, errors='coerce').fillna(0).sum(axis=1)

# Step 2: Feature List
features = [
    'Num_Faults', 'CURRENT_MEAN', 'CURRENT_STD', 'CABLE_AGE',
    'PEAK_STD', 'CYCLE_STD', 'OVR_MEAN', 'OVR_STD',
    'FUSE_COUNT', 'CABLETYPE_RISK', 'MATERIAL_RISK'
]

# Step 3: Filter valid data
df = df.dropna(subset=features)
# Add this before training to generate Health_Score
from datetime import datetime

def compute_health_score(df):
    weights = {
        'Num_Faults': 0.15,
        'CURRENT_MEAN': 0.05,
        'CURRENT_STD': 0.05,
        'CABLE_AGE': 0.15,
        'PEAK_STD': 0.15,
        'CYCLE_STD': 0.05,
        'OVR_MEAN': 0.15,
        'OVR_STD': 0.05,
        'FUSE_COUNT': 0.02,
        'CABLETYPE_RISK': 0.015,
        'MATERIAL_RISK': 0.015
    }

    now_year = datetime.now().year
    if 'DATECREATED' in df.columns:
        years = pd.to_datetime(df['DATECREATED'], errors='coerce').dt.year
        df['CABLE_AGE'] = now_year - years.fillna(now_year)

    current_cols = [c for c in df.columns if c.startswith('I_Month_')]
    if current_cols:
        currents = df[current_cols].apply(pd.to_numeric, errors='coerce').fillna(0)
        df['CURRENT_MEAN'] = currents.mean(axis=1)
        df['CURRENT_STD'] = currents.std(axis=1)

    for prefix, name in [('PEAK_Month_', 'PEAK'), ('CYCLE_Month_', 'CYCLE'), ('OVR_Month_', 'OVR')]:
        cols = [c for c in df.columns if c.startswith(prefix)]
        if cols:
            arr = df[cols].apply(pd.to_numeric, errors='coerce').fillna(0)
            df[f'{name}_MEAN'] = arr.mean(axis=1)
            df[f'{name}_STD'] = arr.std(axis=1)

    if 'CABLETYPE' in df.columns:
        df['CABLETYPE_RISK'] = df['CABLETYPE'].apply(lambda t: 1 if 'PILC' in str(t).upper() else (0.2 if 'XLPE' in str(t).upper() else 0.5))
    if 'CABLECONDUCTORMATERIAL' in df.columns:
        df['MATERIAL_RISK'] = df['CABLECONDUCTORMATERIAL'].apply(lambda m: 0.7 if 'AL' in str(m).upper() else (0.2 if 'CU' in str(m).upper() or 'COPPER' in str(m).upper() else 0.5))

    fuse_cols = [col for col in df.columns if col.startswith('FAULT_RELAY_FUSE')]
    if fuse_cols:
        df['FUSE_COUNT'] = df[fuse_cols].apply(pd.to_numeric, errors='coerce').fillna(0).sum(axis=1)

    used = {k: v for k, v in weights.items() if k in df.columns}
    total_weight = sum(used.values())
    score = pd.Series(0, index=df.index, dtype=float)

    for i in df.index:
        contributions = {}
        for k, w in used.items():
            vals = pd.to_numeric(df[k], errors='coerce').fillna(0)
            rng = vals.max() - vals.min()
            normed = (vals[i] - vals.min()) / rng if rng else 0
            contributions[k] = normed * (w / total_weight)
        score[i] = sum(contributions.values())

    df['Health_Score'] = (1 + 9 * (1 - score)).round().astype(int)
    return df

# Apply the function
df = compute_health_score(df)

# Step 4: Define Target Column
if 'Health_Score' not in df.columns:
    raise ValueError("Missing target column 'Health_Score'. Please compute it or replace with another target.")

X = df[features]
y = df['Health_Score']  # Replace with another column like 'Fault_Prob' if needed

# Step 5: Normalize and Train Model
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

model = RandomForestRegressor(random_state=42)
model.fit(X_scaled, y)

# Step 6: Get Feature Importances
importances = model.feature_importances_

# Step 7: Normalize importances to sum to 1
weight_dict = dict(zip(features, importances))
total = sum(weight_dict.values())
normalized_weights = {k: round(v / total, 4) for k, v in weight_dict.items()}

# Step 8: Print Weights
print("🔍 Learned Feature Weights Based on Data:\n")
for feature, weight in normalized_weights.items():
    print(f"{feature:20} : {weight}")


KeyboardInterrupt: 

EM ALGORITHM BY ADVANCED LOSS FEEDBACK AND IMPROVE THE HEALTH SCORE FUNCTION

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error

def sigmoid(x, k=5):
    return 1 / (1 + np.exp(-k * (x - 0.5)))

# -------------------------------
# 1. Load & Feature Engineering
# -------------------------------
df = pd.read_csv("/media/sagark24/New Volume/MERGE CDIS/IPYNB_FILE/DATA_GENERATION/SWNO_MASTER_COMBINED_FULL.csv")

now_year = pd.Timestamp.now().year
if 'DATECREATED' in df.columns:
    years = pd.to_datetime(df['DATECREATED'], errors='coerce').dt.year
    df['CABLE_AGE'] = now_year - years.fillna(now_year)

current_cols = [c for c in df.columns if c.startswith('I_Month_')]
if current_cols:
    currents = df[current_cols].apply(pd.to_numeric, errors='coerce').fillna(0)
    df['CURRENT_MEAN'] = currents.mean(axis=1)
    df['CURRENT_STD'] = currents.std(axis=1)

for prefix, name in [('PEAK_Month_', 'PEAK'), ('CYCLE_Month_', 'CYCLE'), ('OVR_Month_', 'OVR')]:
    cols = [c for c in df.columns if c.startswith(prefix)]
    if cols:
        arr = df[cols].apply(pd.to_numeric, errors='coerce').fillna(0)
        df[f'{name}_MEAN'] = arr.mean(axis=1)
        df[f'{name}_STD'] = arr.std(axis=1)

if 'CABLETYPE' in df.columns:
    df['CABLETYPE_RISK'] = df['CABLETYPE'].apply(lambda t: 1 if 'PILC' in str(t).upper() else (0.2 if 'XLPE' in str(t).upper() else 0.5))
if 'CABLECONDUCTORMATERIAL' in df.columns:
    df['MATERIAL_RISK'] = df['CABLECONDUCTORMATERIAL'].apply(lambda m: 0.7 if 'AL' in str(m).upper() else (0.2 if 'CU' in str(m).upper() else 0.5))

fuse_cols = [col for col in df.columns if col.startswith('FAULT_RELAY_FUSE')]
if fuse_cols:
    df['FUSE_COUNT'] = df[fuse_cols].apply(pd.to_numeric, errors='coerce').fillna(0).sum(axis=1)

# -------------------------------
# 2. Select Valid Features
# -------------------------------
features = [
    'Num_Faults', 'CURRENT_MEAN', 'CURRENT_STD', 'CABLE_AGE',
    'PEAK_STD', 'CYCLE_STD', 'OVR_MEAN', 'OVR_STD',
    'FUSE_COUNT', 'CABLETYPE_RISK', 'MATERIAL_RISK'
]
df = df.dropna(subset=features)

# Normalize
scaler = MinMaxScaler()
X = pd.DataFrame(scaler.fit_transform(df[features]), columns=features)

# Target
if 'Health_Score' not in df.columns:
    df['Health_Score'] = np.random.randint(1, 10, size=len(df))  # replace this with real scores
y_true = df['Health_Score'].astype(float)

# -------------------------------
# 3. EM-style Learning
# -------------------------------
weights = pd.Series(1 / len(features), index=features)
n_iters = 50
lr = 0.1

for iter in range(n_iters):
    # --- E-Step: Predict Health Score ---
    weighted_sigmoid = (X.apply(sigmoid)).mul(weights).sum(axis=1)
    predicted = 10 * (1 - weighted_sigmoid)
    predicted = predicted.clip(1, 10)

    # --- M-Step: Calculate Error ---
    mse = mean_squared_error(y_true, predicted)
    
    # Compute gradient: smoothed sensitivity
    gradient = {}
    delta = 0.01
    for f in features:
        temp_weights = weights.copy()
        temp_weights[f] += delta
        temp_pred = 10 * (1 - (X.apply(sigmoid)).mul(temp_weights).sum(axis=1))
        temp_mse = mean_squared_error(y_true, temp_pred)
        gradient[f] = (temp_mse - mse) / delta

    # Update weights with regularization (to avoid exploding weights)
    for f in features:
        weights[f] -= lr * gradient[f]

    weights = weights.clip(lower=0)
    weights /= weights.sum()  # normalize

    if iter % 10 == 0 or iter == n_iters - 1:
        print(f"Iteration {iter+1:02d} | MSE: {mse:.4f}")

# -------------------------------
# 4. Final Output
# -------------------------------
print("\n Refined Feature Weights Learned via EM + Feedback:\n")
for f, w in weights.round(4).items():
    print(f"{f:20} : {w}")


Iteration 01 | MSE: 11.9158
Iteration 11 | MSE: 8.8455
Iteration 21 | MSE: 7.0016
Iteration 31 | MSE: 6.8128
Iteration 41 | MSE: 6.8207
Iteration 50 | MSE: 6.7942

✅ Refined Feature Weights Learned via EM + Feedback:

Num_Faults           : 0.0665
CURRENT_MEAN         : 0.1027
CURRENT_STD          : 0.0922
CABLE_AGE            : 0.0392
PEAK_STD             : 0.0475
CYCLE_STD            : 0.0583
OVR_MEAN             : 0.0543
OVR_STD              : 0.0672
FUSE_COUNT           : 0.0443
CABLETYPE_RISK       : 0.0402
MATERIAL_RISK        : 0.3875


ADD A NEW LABELE IS_FAULT

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error

def sigmoid(x, k=5):
    return 1 / (1 + np.exp(-k * (x - 0.5)))

# Load data
df = pd.read_csv("/media/sagark24/New Volume/MERGE CDIS/IPYNB_FILE/DATA_GENERATION/SWNO_MASTER_COMBINED_FULL.csv")

# Feature Engineering
now_year = pd.Timestamp.now().year
if 'DATECREATED' in df.columns:
    years = pd.to_datetime(df['DATECREATED'], errors='coerce').dt.year
    df['CABLE_AGE'] = now_year - years.fillna(now_year)

current_cols = [c for c in df.columns if c.startswith('I_Month_')]
if current_cols:
    currents = df[current_cols].apply(pd.to_numeric, errors='coerce').fillna(0)
    df['CURRENT_MEAN'] = currents.mean(axis=1)
    df['CURRENT_STD'] = currents.std(axis=1)

for prefix, name in [('PEAK_Month_', 'PEAK'), ('CYCLE_Month_', 'CYCLE'), ('OVR_Month_', 'OVR')]:
    cols = [c for c in df.columns if c.startswith(prefix)]
    if cols:
        arr = df[cols].apply(pd.to_numeric, errors='coerce').fillna(0)
        df[f'{name}_MEAN'] = arr.mean(axis=1)
        df[f'{name}_STD'] = arr.std(axis=1)

if 'CABLETYPE' in df.columns:
    df['CABLETYPE_RISK'] = df['CABLETYPE'].apply(lambda t: 1 if 'PILC' in str(t).upper() else (0.2 if 'XLPE' in str(t).upper() else 0.5))

if 'CABLECONDUCTORMATERIAL' in df.columns:
    df['MATERIAL_RISK'] = df['CABLECONDUCTORMATERIAL'].apply(lambda m: 0.7 if 'AL' in str(m).upper() else (0.2 if 'CU' in str(m).upper() else 0.5))

fuse_cols = [col for col in df.columns if col.startswith('FAULT_RELAY_FUSE')]
if fuse_cols:
    df['FUSE_COUNT'] = df[fuse_cols].apply(pd.to_numeric, errors='coerce').fillna(0).sum(axis=1)

# Features & target
features = [
    'Num_Faults', 'CURRENT_MEAN', 'CURRENT_STD', 'CABLE_AGE',
    'PEAK_STD', 'CYCLE_STD', 'OVR_MEAN', 'OVR_STD',
    'FUSE_COUNT', 'CABLETYPE_RISK', 'MATERIAL_RISK'
]
df = df.dropna(subset=features)

X = pd.DataFrame(MinMaxScaler().fit_transform(df[features]), columns=features)

# REAL TARGET — use actual number of faults
# Option A: Use actual count (recommended for regression)
y_true = df['Num_Faults'].clip(upper=10).astype(float)

# Option B: Use binary target (for classification)
# y_true = (df['Num_Faults'] > 0).astype(float)

# Feature penalty boosts (domain knowledge)
penalty_boosts = {
    'Num_Faults': 2.0,
    'CABLE_AGE': 1.5,
    'MATERIAL_RISK': 1.2,
    'CURRENT_MEAN': 1.0,
    'CURRENT_STD': 1.0,
    'PEAK_STD': 1.0,
    'CYCLE_STD': 1.0,
    'OVR_MEAN': 1.0,
    'OVR_STD': 1.0,
    'FUSE_COUNT': 1.0,
    'CABLETYPE_RISK': 1.0
}

# EM-style training
weights = pd.Series(1 / len(features), index=features)
n_iters = 500
lr = 0.1

for iter in range(n_iters):
    weighted_sigmoid = (X.apply(sigmoid)).mul(weights).sum(axis=1)
    predicted = 10 * (1 - weighted_sigmoid)
    predicted = predicted.clip(0, 10)

    mse = mean_squared_error(y_true, predicted)

    gradient = {}
    delta = 0.01
    for f in features:
        temp_weights = weights.copy()
        temp_weights[f] += delta
        temp_pred = 10 * (1 - (X.apply(sigmoid)).mul(temp_weights).sum(axis=1))
        temp_mse = mean_squared_error(y_true, temp_pred)
        grad = (temp_mse - mse) / delta

        # Apply feature-specific penalty
        gradient[f] = grad * penalty_boosts.get(f, 1.0)

    for f in features:
        weights[f] -= lr * gradient[f]

    weights = weights.clip(lower=0)
    weights /= weights.sum()

    if iter % 10 == 0 or iter == n_iters - 1:
        print(f"Iteration {iter+1:02d} | MSE: {mse:.4f}")

# Final output
print("\nFinal Weights (EM + Real Target + Penalized Updates):\n")
for f, w in weights.round(4).items():
    print(f"{f:20} : {w}")


Iteration 01 | MSE: 48.7604
Iteration 11 | MSE: 20.1903
Iteration 21 | MSE: 20.1903
Iteration 31 | MSE: 20.1903
Iteration 41 | MSE: 20.1903
Iteration 51 | MSE: 20.1903
Iteration 61 | MSE: 20.1903
Iteration 71 | MSE: 20.1903
Iteration 81 | MSE: 20.1903
Iteration 91 | MSE: 20.1903
Iteration 101 | MSE: 20.1903
Iteration 111 | MSE: 20.1903
Iteration 121 | MSE: 20.1903
Iteration 131 | MSE: 20.1903
Iteration 141 | MSE: 20.1903
Iteration 151 | MSE: 20.1903
Iteration 161 | MSE: 20.1903
Iteration 171 | MSE: 20.1903
Iteration 181 | MSE: 20.1903
Iteration 191 | MSE: 20.1903
Iteration 201 | MSE: 20.1903
Iteration 211 | MSE: 20.1903
Iteration 221 | MSE: 20.1903
Iteration 231 | MSE: 20.1903
Iteration 241 | MSE: 20.1903
Iteration 251 | MSE: 20.1903
Iteration 261 | MSE: 20.1903
Iteration 271 | MSE: 20.1903
Iteration 281 | MSE: 20.1903
Iteration 291 | MSE: 20.1903
Iteration 301 | MSE: 20.1903
Iteration 311 | MSE: 20.1903
Iteration 321 | MSE: 20.1903
Iteration 331 | MSE: 20.1903
Iteration 341 | MSE: 20.

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error

# Sigmoid function for health degradation scaling
def sigmoid(x, k=5):
    return 1 / (1 + np.exp(-k * (x - 0.5)))

# Load and clean data
df = pd.read_csv("/media/sagark24/New Volume/MERGE CDIS/IPYNB_FILE/DATA_GENERATION/SWNO_MASTER_COMBINED_FULL.csv")

# Feature Engineering
now_year = pd.Timestamp.now().year
if 'DATECREATED' in df.columns:
    years = pd.to_datetime(df['DATECREATED'], errors='coerce').dt.year
    df['CABLE_AGE'] = now_year - years.fillna(now_year)

current_cols = [c for c in df.columns if c.startswith('I_Month_')]
if current_cols:
    current = df[current_cols].apply(pd.to_numeric, errors='coerce').fillna(0)
    df['CURRENT_MEAN'] = current.mean(axis=1)
    df['CURRENT_STD'] = current.std(axis=1)

for prefix, name in [('PEAK_Month_', 'PEAK'), ('CYCLE_Month_', 'CYCLE'), ('OVR_Month_', 'OVR')]:
    cols = [c for c in df.columns if c.startswith(prefix)]
    if cols:
        arr = df[cols].apply(pd.to_numeric, errors='coerce').fillna(0)
        df[f'{name}_MEAN'] = arr.mean(axis=1)
        df[f'{name}_STD'] = arr.std(axis=1)

if 'CABLETYPE' in df.columns:
    df['CABLETYPE_RISK'] = df['CABLETYPE'].apply(lambda t: 1 if 'PILC' in str(t).upper() else (0.2 if 'XLPE' in str(t).upper() else 0.5))

if 'CABLECONDUCTORMATERIAL' in df.columns:
    df['MATERIAL_RISK'] = df['CABLECONDUCTORMATERIAL'].apply(lambda m: 0.7 if 'AL' in str(m).upper() else (0.2 if 'CU' in str(m).upper() else 0.5))

fuse_cols = [c for c in df.columns if c.startswith('FAULT_RELAY_FUSE')]
if fuse_cols:
    df['FUSE_COUNT'] = df[fuse_cols].apply(pd.to_numeric, errors='coerce').fillna(0).sum(axis=1)

# Select relevant features
features = [
    'Num_Faults', 'CURRENT_MEAN', 'CURRENT_STD', 'CABLE_AGE',
    'PEAK_STD', 'CYCLE_STD', 'OVR_MEAN', 'OVR_STD',
    'FUSE_COUNT', 'CABLETYPE_RISK', 'MATERIAL_RISK'
]

# Drop incomplete rows
df = df.dropna(subset=features)

# Normalize feature values
X = pd.DataFrame(MinMaxScaler().fit_transform(df[features]), columns=features)

# Ground truth: fault count (or clip to max 10)
y_true = df['Num_Faults'].clip(upper=10).astype(float)

# Domain-based penalty scaling per feature
penalty_boosts = {
    'Num_Faults': 2.0,
    'CABLE_AGE': 1.5,
    'MATERIAL_RISK': 1.2,
    'CURRENT_MEAN': 1.0,
    'CURRENT_STD': 1.0,
    'PEAK_STD': 1.0,
    'CYCLE_STD': 1.0,
    'OVR_MEAN': 1.0,
    'OVR_STD': 1.0,
    'FUSE_COUNT': 1.0,
    'CABLETYPE_RISK': 1.0
}

# Initialize weights uniformly
weights = pd.Series(1 / len(features), index=features)

# Normalize y_true to [0, 1] for stability
y_true_scaled = y_true / 10.0

# EM-style optimization with regularization
n_iters = 100
lr = 0.01
l2_penalty = 0.01  # L2 regularization strength

for iter in range(n_iters):
    sigmoid_vals = X.apply(sigmoid)
    predicted = (1 - sigmoid_vals.mul(weights).sum(axis=1)).clip(0, 1)
    mse = mean_squared_error(y_true_scaled, predicted)

    gradient = {}
    delta = 1e-4
    for f in features:
        temp_weights = weights.copy()
        temp_weights[f] += delta
        temp_weights /= temp_weights.sum()

        temp_pred = (1 - X.apply(sigmoid).mul(temp_weights).sum(axis=1)).clip(0, 1)
        temp_mse = mean_squared_error(y_true_scaled, temp_pred)
        grad = (temp_mse - mse) / delta

        # Add L2 regularization
        reg_term = 2 * weights[f] * l2_penalty
        gradient[f] = (grad + reg_term) * penalty_boosts.get(f, 1.0)

    # Update and normalize
    for f in features:
        weights[f] -= lr * gradient[f]

    weights = weights.clip(lower=1e-4)  # prevent collapse
    weights /= weights.sum()

    if iter % 10 == 0 or iter == n_iters - 1:
        print(f"Iteration {iter+1:03d} | MSE: {mse:.4f}")


# Print learned weights
print("\n Final Weights (EM + Real Target + Penalized Updates):\n")
for f, w in weights.round(4).items():
    print(f"{f:20} : {w}")

# Final Health Score using learned weights
df['Advanced_Health_Score'] = (10 * (1 - X.apply(sigmoid).mul(weights).sum(axis=1))).clip(0, 10).round()

# Save if needed
# df.to_csv("final_health_score_output.csv", index=False)


Iteration 001 | MSE: 0.4876
Iteration 011 | MSE: 0.3405
Iteration 021 | MSE: 0.2280
Iteration 031 | MSE: 0.1424
Iteration 041 | MSE: 0.1050
Iteration 051 | MSE: 0.0971
Iteration 061 | MSE: 0.0903
Iteration 071 | MSE: 0.0838
Iteration 081 | MSE: 0.0777
Iteration 091 | MSE: 0.0719
Iteration 100 | MSE: 0.0671

✅ Final Weights (EM + Real Target + Penalized Updates):

Num_Faults           : 0.0001
CURRENT_MEAN         : 0.1439
CURRENT_STD          : 0.0001
CABLE_AGE            : 0.1667
PEAK_STD             : 0.0001
CYCLE_STD            : 0.0531
OVR_MEAN             : 0.0001
OVR_STD              : 0.0001
FUSE_COUNT           : 0.0001
CABLETYPE_RISK       : 0.0001
MATERIAL_RISK        : 0.6356
